### PR TITLE
Feature #5980: proposing a new space and component navigation in order to improve display performances when it exists a lot of spaces and subspaces.

### PR DESCRIPTION
**Pull request by SilverYoCha**

### CODE REVIEW

File: core-war/src/main/java/org/silverpeas/web/jobstartpage/servlets/JobStartPagePeasRequestRouter.java [hunk 1]
```diff
@@ -752,6 +768,71 @@ public String getDestination(String function, JobStartPagePeasSessionController
...
+            }
+            o.putJSONArray("path", a -> encodeSortedPath.apply(path.stream(), a));
+          }
+          o.putJSONArray("spaces", a -> encodeSortedSubSpaces.apply(jobStartPageSC.getSubSpaces(), a));
+          o.putJSONArray("applications", a -> encodeSortedInstances.apply(
+              isDefined(jobStartPageSC.getSubSpaceId()) ?
+                  jobStartPageSC.getSubSpaceComponents() :
+                  jobStartPageSC.getSpaceComponents(), a));
+          return o;
+        }));
```
Comment (mmoqui) [inline core-war/src/main/java/org/silverpeas/web/jobstartpage/servlets/JobStartPagePeasRequestRouter.java:835]: Hum... this JSON construction is a complex code (in term of complexity level). Could it be possible to to have a POJO to represent such a JSON object, to build it with a builder, and then, by using `JSONCodec.encode(T)`, to convert it into a JSON `String` representation?

File: core-war/src/main/webapp/jobStartPagePeas/jsp/javascript/vuejs/admin-navigation.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                selectorCurrentSpace : undefined,
+                path : undefined,
+                spaces : undefined,
+                applications : undefined
+              }, data);
+              this.selectorSpaces = data.selectorSpaces;
+              this.selectorCurrentSpace = data.selectorCurrentSpace;
+              const currentTreeLevel = [];
+              currentTreeLevel.path = data.path;
+              Array.prototype.push.apply(currentTreeLevel, data.spaces);
```
Comment (mmoqui) [inline core-war/src/main/webapp/jobStartPagePeas/jsp/javascript/vuejs/admin-navigation.js:58]: Why don't call directly `currentTreeLevel.push(...)` method as we now here `currentTreeLevel` is well an array?

File: core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java [hunk 3]
```diff
@@ -23,57 +23,54 @@
...
     DisplaySorted elmt = getSpaceCache(spaceId);
     if (elmt != null) {
       elmt.copy(buildSpaceObject(spaceId));
-      if (spaceId.equals(m_CurrentSpaceId)) {
-        setCurrentSpace(m_CurrentSpaceId);
-      } else if (spaceId.equals(m_CurrentSubSpaceId)) {
+      if (spaceId.equals(currentSpaceId)) {
+        setCurrentSpace(currentSpaceId);
+      } else if (spaceId.equals(currentSubSpaceId)) {
         setCurrentSubSpace(null);
```
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java:74]: Yes indeed, there is no point to set null before.

File: core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java [hunk 4]
```diff
@@ -82,107 +79,91 @@ public void resetSpaceCache(String theSpaceId) {
...
+    spContext = URLUtil.getApplicationURL();
+    adminCtrl = ServiceProvider.getService(AdminController.class);
+    sessionCtrl = msc;
+    this.user = user;
+    elCounter = 0;
+    currentSpaceId = null;
+    currentSubSpaceId = null;
+    subSpaces = new DisplaySortedCache();
+    spaceComponents = new DisplaySortedCache();
+    subSpaceComponents = new DisplaySortedCache();
```
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java:144]: Yes, I will implement the clear() sugnature.

File: core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java [hunk 5]
```diff
@@ -368,174 +308,72 @@ protected String getShortSpaceId(String spaceId) {
...
+          ds.setType(DisplaySorted.TYPE_COMPONENT);
+          ds.setTypeName(ci.isWorkflow() ? "processManager" : ci.getName());
+          ds.setAdmin(isTheSpaceAdmin);
+          ds.setDeep(spaceInst.getLevel());
+          ds.setVisible(isTheSpaceAdmin);
+          return ds;
+        });
+  }
+
+  private static class DisplaySortedCache {
```
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java:344]: :thumbsup:

File: core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java [hunk 6]
```diff
@@ -82,107 +79,91 @@ public void resetSpaceCache(String theSpaceId) {
...
       for (String manageableSpaceId : allManageableSpaceIds) {
-        m_ManageableSpaces.add(getShortSpaceId(manageableSpaceId));
+        manageableSpaces.add(getShortSpaceId(manageableSpaceId));
       }
     }
 
-    String[] spaceIds = m_administrationCtrl.getAllRootSpaceIds();
-    m_Spaces = createSpaceObjects(spaceIds, false);
+    String[] spaceIds = adminCtrl.getAllRootSpaceIds();
+    spaces = new DisplaySortedCache(createSpaceObjects(stream(spaceIds), false));
```
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/jobstartpage/NavBarManager.java:156]: :thumbsup:

File: core-war/src/main/java/org/silverpeas/web/jobstartpage/control/JobStartPagePeasSessionController.java [hunk 7]
```diff
@@ -166,9 +170,9 @@ public boolean isManagedSpaceRoot() {
 
   public DisplaySorted[] getManagedSpaceComponents() {
```
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/jobstartpage/control/JobStartPagePeasSessionController.java:171]: Changing this signature has impacts on existing code which I do not want to spend time on it.

File: core-war/src/main/java/org/silverpeas/web/jobstartpage/servlets/JobStartPagePeasRequestRouter.java (hunk 1 above)
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/jobstartpage/servlets/JobStartPagePeasRequestRouter.java:835]: OK, I will export this JSON encoding in a dedicated encoder.

File: core-war/src/main/webapp/jobStartPagePeas/jsp/javascript/vuejs/admin-navigation.js (hunk 2 above)
Comment (SilverYoCha) [reply core-war/src/main/webapp/jobStartPagePeas/jsp/javascript/vuejs/admin-navigation.js:58]: After some inconveniences in the past, I took the habit to do that way.

**Review by mmoqui — APPROVED** [review]
All is ok

**Pull request merged**